README.md
# Videk client

Videk client and related boot scripts. This package provides Videk REST client
and a systemd service for registration of SNA-LGTC boards in the Videk management
system.

To build Debian package:

    $ dpkg-buildpackage -b -uc

When built, install the package with:

    $ sudo dpkg -i videk-client_1.1.0-1_all.deb

To add a new release in change log:

    $ dch -i

After installation, the configuration is in `/etc/videk/conf` file. Videk API key must be written under key section. This key can be found at videk.ijs.si by selecting
*Users*, *admin*. The API key is displayed under *API Token* header under *User
info*.

If you are installing this package on Raspberry pi in `/etc/videk/conf` file set:

    $ id=rpi-serial


debian/changelog
videk-client (3.4.0-1) unstable; urgency=medium

  * Fix HTTP session hang

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 09 Dec 2020 10:56:08 +0100

videk-client (3.3.0-1) unstable; urgency=medium

  * Fix same host bug again

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 30 Sep 2020 14:41:19 +0200

videk-client (3.2.0-1) unstable; urgency=medium

  * Fix same host bug

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Mon, 28 Sep 2020 14:29:01 +0200

videk-client (3.1.0-1) unstable; urgency=medium

  * Add grideye support 

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 17 Jun 2020 21:19:01 +0200

videk-client (3.0.0-1) unstable; urgency=medium

  * Add ebottle support

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 06 May 2020 14:38:23 +0200

videk-client (2.1.0-1) unstable; urgency=medium

  * Add after network online to systemd service

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 14 Nov 2018 13:19:04 +0100

videk-client (2.0.0-1) unstable; urgency=medium

  * Add complete videk-client setup.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Tue, 11 Sep 2018 14:02:50 +0200

videk-client (1.4.0-1) unstable; urgency=medium

  * Add temporary node name change.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Mon, 07 May 2018 11:00:45 +0200

videk-client (1.3.0-1) unstable; urgency=medium

  * Add external conf file.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Tue, 16 Jan 2018 12:35:22 +0100

videk-client (1.2.0-1) unstable; urgency=medium

  * Add periodic hostname update support.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Thu, 11 Jan 2018 09:37:24 +0100

videk-client (1.1.0-1) unstable; urgency=medium

  * Add staus update support.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Wed, 06 Sep 2017 13:13:14 +0200

videk-client (1.0.0-1) unstable; urgency=medium

  * Initial release.

 -- Matevz Vucnik <matevz.vucnik@ijs.si>  Tue, 22 Aug 2017 09:47:13 +0200


videk-ebottle.py
#!/usr/bin/env python3

import re
import sys
import json
import time
import socket
import subprocess
import netifaces as ni
from datetime import datetime
from videkrestclient import Videk

lat = 46.042767
lon = 14.487632

def getserial():
    cpuserial = "0000000000000000"
    try:
        f = open('/proc/cpuinfo','r')
        for line in f:
            if line[0:6]=='Serial':
                cpuserial = line[10:26]
        f.close()
    except:
        cpuserial = "ERROR000000000"
 
    return cpuserial

conf = {}
with open('conf') as confile:
    for line in confile:
        name, var = line.partition("=")[::2]
        conf[name.strip()] = var.strip()

videk_api_key = conf['key']
videk = Videk(conf['api'], videk_api_key)

current_ip = None

if conf['id'] == "rpi-serial":
    machine_id = getserial()
else:
    machine_id = open('/etc/machine-id').readline().strip()

if videk.serverOnline():
    print("Videk server is online ...")
else:
    print("Videk server is offline ...")
    sys.exit(1)

def uploadSensors(node, node_id, sensor_type, sensors):
    for sensor in sensors:        
        sensor_id = videk.getSensorID(node, sensor_type, sensor['name'])
        if sensor_id == None:
            videk.createSensor(node_id, sensor_type, sensor['name'],
                sensor['unit'])
            sensor_id = videk.getSensorID(node, sensor_type, sensor['name'])

        measurement = '''{"latitude":"","longitude":"","ts":"","value":""}'''
        v = sensor['value']
        preparedData = []
        data = json.loads(measurement)
        data['value'] = v
        data['ts'] = datetime.utcnow().isoformat()
        data['latitude'] = lat
        data['longitude'] = lon
        preparedData.append(data)

        videk.uploadMesurements(preparedData, node_id, sensor_id)

def readSensors():
    sensors = []

    try:
        with open('sensors') as sensor_file:
            for line in sensor_file:
                sensor = {}
                name, var = line.partition("=")[::2]

                sensor['name'] = name.strip()
                sensor['value'] = round(float(var.strip()), 2)

                if 'temperature' in name:
                    sensor['unit'] = "C"
                elif 'battery' in name:
                    sensor['unit'] = "%"
                elif 'luminance' in name:
                    sensor['unit'] = "lm"
                elif 'position' in name:
                    sensor['unit'] = "G"
                
                sensors.append(sensor)
    except:
        pass

    return sensors

def writeLeds(extra_fields):
    for el in extra_fields:
        try:
            leds = el['LEDs']
            f = open("/tmp/ebottle/leds", "w")
            f.write(leds + "\n")
            f.close()
        except:
            pass

while True:

    node_model = videk.getNodeByHardwareId(machine_id)
    node_name = node_model['name']
    node_id = int(node_model['id'])

    #writeLeds(node_model['extra_fields'])
    
    s = readSensors()
    
    uploadSensors(node_name, node_id, "ebottle", s)

    time.sleep(10)


videkrestclient/__init__.py
import sys
import json
import requests

class Videk:
    url = "http://localhost:3000"
    api_url = "/api"
    nodes_url = "/nodes"
    sensors_url = "/sensors"
    clusters_url = "/clusters"
    measurements_url = "/measurements"
    latitude = "null"
    longitude = "null"
    token = ""
    headers = {'Content-Type': 'application/json', 'Authorization': ''}
    session = requests.Session()

    def __init__(self, url, token):
        self.url = url
        self.api_url = self.url + self.api_url
        self.token = token
        self.headers = {'Content-Type': 'application/json', 'Authorization': token}

    def createCluster(self, clusterName, clusterType):
        json_str = '''{ "name": "''' + clusterName + '''", "id": "''' + clusterName + '''", "tag": null, "type": "''' + clusterType + '''", "URL": null, "scan" : "false", "comment":""  }'''
        try:
            json_str = json_str.encode('utf8')
            r = self.session.post(self.api_url + self.clusters_url, data=json_str, headers=self.headers)
            print(r.text)
            if "error" in r.text:
                print("Error: Cluster with the name " + clusterName + " already exists")
                sys.exit()
        except requests.exceptions.RequestException as e:
            print(e)

    def getClusterID(self, clusterName):
        try:
            r = self.session.get(self.api_url + self.clusters_url + "?name=" + clusterName, headers=self.headers)
            cluster_id = r.json()
            if len(cluster_id) == 0:
                print("Error: Cluster with the name " + clusterName + " not found")
            else:
                return cluster_id[0]['id']
        except requests.exceptions.RequestException as e:
            print(e)

    def getClusterName(self, clusterID):
        try:
            r = self.session.get(self.api_url + self.clusters_url + "?id=" + clusterID, headers=self.headers)
            cluster_name = r.json()
            if len(cluster_name) == 0:
                print("Error: Cluster with the id " + clusterID + " not found")
            else:
                return cluster_name[0]['name']
        except requests.exceptions.RequestException as e:
            print(e)

    def createNode(self, nodeName, clusterID):
        try:
            clusterName = self.getClusterName(clusterID)
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + nodeName, headers=self.headers)
            if "No nodes found" in r.text:
                json_str = '''{ "name": "''' + nodeName + '''", "loc_lat": ''' + str(self.latitude) + ''',
                "loc_lon": ''' + str(self.longitude) + ''', "cluster": "''' + str(clusterID) + '''",
                "cluster_name": "''' + clusterName + '''", "status": "active", "components": [] }'''
                json_str = json_str.encode('utf8')
                r = self.session.post(self.api_url + self.nodes_url, data=json_str, headers=self.headers)
                print(r.text)
                if "error" in r.text:
                    print("Error: Node with the name  " + nodeName + " already exists")
                    sys.exit()
        except requests.exceptions.RequestException as e:
            print(e)

    def updateNode(self, node_id, nodeModel):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?id=" + str(node_id), headers=self.headers)
            res = r.json()
            if "No nodes found." in res:
                print(res)
            else:
                node_id = res[0]['_id']
                r = self.session.put(self.api_url + self.nodes_url + "/" + str(node_id), data=json.dumps(nodeModel), headers=self.headers)
                print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def updateSingleNodeParam(self, nodeName, key, value):
        modData = { key: value }
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?id=" + str(nodeName), headers=self.headers)
            res = r.json()
            if "No nodes found." in res:
                print(res)
            else:
                node_id = res[0]['_id']
                r = self.session.put(self.api_url + self.nodes_url + "/" + str(node_id), data=json.dumps(modData), headers=self.headers)
                print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def addNodeExtraField(self, nodeName, fieldName, fieldValue):
        node_model = self.getNode(str(nodeName))
        if 'extra_fields' in node_model:
            node_model_update = node_model['extra_fields']
            node_model_update.append({ fieldName: fieldValue })
            self.updateSingleNodeParam(node_model['id'], "extra_fields", node_model_update)
        else:
            self.updateSingleNodeParam(node_model['id'], "extra_fields", [{ fieldName: fieldValue }])

    def getNodeID(self, nodeName):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + nodeName, headers=self.headers)
            node_id = r.json()
            if len(node_id) == 15:
                print("Error: Node with the name " + nodeName + " not found")
            else:
                return node_id[0]['id']
        except requests.exceptions.RequestException as e:
            print(e)

    def getNodeByHardwareId(self, id):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?machine_id=" + id, headers=self.headers)
            res = r.json()
            if len(res) == 15:
                print("Error: Node with the machine ID " + id + " not found")
            else:
                node_id = res[0]['_id']
                node = self.session.get(self.api_url + self.nodes_url + "/" + str(node_id), headers=self.headers).json()
                return node[0]
        except requests.exceptions.RequestException as e:
            print(e)

    def getNode(self, id):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + id, headers=self.headers)
            res = r.json()
            if len(res) == 15:
                print("Error: Node with the name " + id + " not found")
            else:
                node_id = res[0]['_id']
                node = self.session.get(self.api_url + self.nodes_url + "/" + str(node_id), headers=self.headers).json()
                return node[0]
        except requests.exceptions.RequestException as e:
            print(e)

    def getNodeLocation(self, nodeName):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + nodeName, headers=self.headers)
            node_id = r.json()
            if len(node_id) == 15:
                print("Error: Node with the name " + nodeName + " not found")
            else:
                node_id = node_id[0]['_id']
                node = self.session.get(self.api_url + self.nodes_url + "/" + str(node_id), headers=self.headers).json()
                return { "latitude": float(node[0]['loc_lat']), "longitude": float(node[0]['loc_lon']) }
        except requests.exceptions.RequestException as e:
            print(e)

    def getNodeName(self, nodeID):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?id=" + nodeID, headers=self.headers)
            node_name = r.json()
            if len(node_name) == 0:
                print("Error: Node with the id " + nodeID + " not found")
            else:
                return node_name[0]['name']
        except requests.exceptions.RequestException as e:
            print(e)

    def createSensor(self, nodeID, sensorType, sensorQuantity, sensorUnit):
        try:
            sensor_id = str(nodeID) + "-" + sensorType + "-" + sensorQuantity
            r = self.session.get(self.api_url + self.sensors_url + "?id=" + sensor_id, headers=self.headers)
            if "No sensors found" in r.text:
                r = self.session.get(self.api_url + self.nodes_url + "?id=" + str(nodeID), headers=self.headers)
                node_id = r.json()[0]['_id']
                json_str = '''{ "id": "''' + sensor_id + '''", "type": "''' + sensorType + '''", "quantity":
                "''' + sensorQuantity + '''", "unit": "''' + sensorUnit + '''", "node_id": "''' + node_id + '''",
                "node": ''' + str(nodeID) + ''' }'''
                json_str = json_str.encode('utf8')
                r = self.session.post(self.api_url+self.sensors_url, data=json_str, headers=self.headers)
                print(r.text)
            else:
                print("Sensor already exists")
        except requests.exceptions.RequestException as e:
            print(e)

    def getSensorID(self, nodeName, sensorType, sensorQuantity):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + nodeName, headers=self.headers)
            data = r.json()
            node_id = data[0]['id']
            sensor_id = str(node_id) + "-" + sensorType + "-" + sensorQuantity
            r = self.session.get(self.api_url + self.sensors_url + "?id=" + str(sensor_id), headers=self.headers)
            if "No sensors found" in r.text:
                print("No sensors found")
            else:
                return sensor_id
        except requests.exceptions.RequestException as e:
            print(e)

    def uploadMesurements(self, mesurements, nodeID, sensorID):
        try:
            r = self.session.get(self.api_url + self.sensors_url + "?id=" + sensorID, headers=self.headers)
            sensor_mongo_id = r.json()[0]['_id']
            r = self.session.get(self.api_url + self.nodes_url + "?id=" + str(nodeID), headers=self.headers)
            node_mongo_id = r.json()[0]['_id']
            measurement = '''{"sensor_id":"''' + sensor_mongo_id + '''","node_id": "''' + node_mongo_id + '''","sensor":
            "''' + sensorID + '''","node": ''' + str(nodeID) + ''', "latitude": 99.999999 , "longitude": 99.999999 ,"ts":
             "2014-07-24T15:14:30.850Z","value": 0,"context":"data"}'''
            preparedData = []
            for x in mesurements:
                data = json.loads(measurement)
                data['value'] = x['value']
                data['ts'] = x['ts']
                data['latitude'] = x['latitude']
                data['longitude'] = x['longitude']
                preparedData.append(data)
            r = self.session.post(self.api_url+self.measurements_url, data=json.dumps(preparedData), headers=self.headers)
            print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def deleteCluster(self, clusterName):
        try:
            r = self.session.get(self.api_url + self.clusters_url + "?name=" + clusterName, headers=self.headers)
            data = r.json()
            if len(data) == 0:
                print("No sensors found.")
            else:
                cluster_id = data[0]['_id']
                r = self.session.delete(self.api_url + self.clusters_url + "/" + str(cluster_id), headers=self.headers)
                print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def deleteNode(self, nodeName):
        try:
            r = self.session.get(self.api_url + self.nodes_url + "?name=" + nodeName, headers=self.headers)
            data = r.json()
            if "No nodes found." in data:
                print(data)
            else:
                node_id = data[0]['_id']
                r = self.session.delete(self.api_url + self.nodes_url + "/" + str(node_id), headers=self.headers)
                print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def deleteSensor(self, nodeID, sensorType, sensorQuantity):
        try:
            sensor_id = str(nodeID) + "-" + sensorType + "-" + sensorQuantity
            r = self.session.get(self.api_url + self.sensors_url + "?id=" + sensor_id, headers=self.headers)
            if "No sensors found" in r.text:
                print(r.text)
            else:
                data = r.json()
                sensor_id = data[0]['_id']
                r = self.session.delete(self.api_url + self.sensors_url + "/" + str(sensor_id), headers=self.headers)
                print(r.text)
        except requests.exceptions.RequestException as e:
            print(e)

    def serverOnline(self):
        try:
            request = self.session.get(self.url)
        except:
            return False
            pass
        if request.status_code == 200:
            return True
        else:
            return False
